Obstacle system refreshing

Illithidek opened the pull request:
I want redesign/add some obstacles functionality. Like any wip, it's far away from end.
Anyway, I put it here for all to know that such thing exists.
Pls, don't comment about code formatting, serialization corruption with old saves, some unused code etc. In few words for me there are things that I will do during work on it.

EDIT:
TODO for now:

- [x] Primary obstacles

- [x] Basic modding for static obstacles 

- [ ] Basic modding for magic obstacles

- [x] Suggested by _Andruids_: Adding dissapear and intereaction animation

Commit:
Obstacle rebasing

Better managing obstacle area

Create a new class that will handling obstacle area, fixing arithmetic
of
old system for moving area to some position depending on the old
position, also now absolute obstacle can be moved like usual.
Making unit test for that class.

Move ObstacleType to separating header

From CObstacleInstance.h to ObstacleType.h.
Change to strong typed enum.

Creating ObstacleSurface class

Simple moving some objects and change them name from ObstacleInfo to
ObstacleSurface.
Making unit test for new class.

Changing "ObstacleInfo" class

Now ObstacleInfo is used for getting information that describe obstacle
from specified json.

Using ObstacleType from ObstacleType.h

Delete old class-scope ObstacleType enum, and use new one on it's place
- strong enum.

Handling new classes by old code

Handling these new classes:
ObstacleArea,ObstacleSurface,ObstacleInfo
and delete old implementation usage

Delete all "id" fields in obstacles.json

Delete all "id" fields in config/obstacles.json since they are unused

Rid obstacles from CHeroHandler

Remove obstacles methods, variables etc from CHeroHandler.
Improve some methods, and so on more actions for improve the obstacle
code.

Further improvements for obstacles

Now should be possible moving for example all moat by some offset.
Also now we can delete some fields from obstacle area and add it
dynamically during the game.

Obstacle code cleanup.

Automation of counting

Automation of counting width and height for usual obstacles.

Absolute and obstacles code blit connection.

Connect two separate parts of code into one:
Absolute obstacles blit and obstacles blit.

Move CObstacleInstance to "obstacle" subdirectory

Rename ObstacleInfo to ObstacleJson

Obstacle speed optimization

Loading from json data every time when refreshing render takes up a lot
of memory, so it's better to store objects in a couple of variables, and
initialize them one time at the battle beginning.

Setting area for spell obstacles in obstacleMechanics class.

Obstacle code improvements

Fix in force field graphics shift.
Delete "ID" variable and change "uniqueID" name to "ID".
Delete "ABSOLUTE_OBSTACLE" type from code

Some improvements in obstacle code

-Change USUAL type to STATIC in ObstacleType.h
-Remove getting graphics from effectID for obstacles created by spell.
-Merge absoluteobstacles and obstacles into one table in obstacles.json.
-Making flag in obstacle json, for knowing which static obstacles can be
removed by spell.
- Rename "defName" to "graphicsName" in obstacle code.

Remove converting obstacles BattlefieldType to BattlefieldBI

Obstacles new features

In general:
- Change rendering things.
- Big changing in settings proper obstacles inside json, now it's
possible to set most things by him.
- Splitting initialization part of obstacle code into 2 systems, first
part that will generate, random obstacles(not yet finished part), and
second that will handle inherent obstacles.
- Support for handling towns and creature banks.

Cleanup in obstacle code, separate some files, making few tests.

Serialization backward-compatibility fix

Fixing some minor problems with obstacle code.

Delete "moat" from config/schemas/townSiege.json

Support more types for randomless obstacles.

Making simple handler for battlefield.

-Battlefield handler now supports obstacles mods.
-Change obstacles.json construction from vector to structure.
-Added empty schema for requirements

Making BattlefieldType converters

-rename BFieldType to BattlefieldType.
-adding method that convert battlefieldType toString() and fromString()
-adding tests for BattlefieldType converters

Improvement for naming in obstacles.json

Replace integers to strings
Create BATTLEFIELD_TYPE_NAMES in StringConstants.h
Small fix in battleBackgrounds associated with the removal of
EBattlefieldType::NONE2 type.

-Delete ObstacleSideOptions in Obstacle effect
-Use ObstacleArea and ObstacleGraphicsInfo instead ObstacleSideOptions
-Adjustment spell json
-Use SpellID in obstacles
-Adding "consistent" parameter in obstacle effect

ArseniyShestakov commented:
Please consider fixing this as well:
https://bugs.vcmi.eu/view.php?id=2786

FeniksFire commented:
Fixed: https://bugs.vcmi.eu/view.php?id=2786

alexvins commented on lib/battle/obstacle/ObstacleJson.h:
```diff
@@ -10,21 +10,37 @@
+	virtual int32_t getDamage() const;
+
+	virtual int8_t getSpellLevel() const;
+	virtual int8_t getSpellPower() const;
```
Spell power should be 32 bit

alexvins commented on lib/battle/obstacle/ObstacleJson.h:
```diff
@@ -10,21 +10,37 @@
+	virtual int8_t getSpellPower() const;
+	virtual int8_t getBattleSide() const;
+	virtual int8_t isVisibleForAnotherSide() const;
+	virtual int8_t getTurnsRemaining() const;
```
Turns remaining should be 16 bit at least

FeniksFire commented:
Serialization for old saves should working now.

alexvins commented on lib/NetPacks.h:
```diff
@@ -1738,6 +1738,11 @@ struct ObstaclesRemoved : public CPackForClient
 
 	template <typename Handler> void serialize(Handler &h, const int version)
 	{
+		if(version < 777 && !h.saving)
```
We do not use backward compatibility for NetPacks, server and client must have same version.

alexvins commented on lib/CTownHandler.h:
```diff
@@ -253,6 +253,21 @@ class DLL_LINKAGE CTown
+		}
+		else if(!h.saving)
+		{
+			std::vector<BattleHex> moatHexes = {11, 28, 44, 61, 77, 111, 129, 146, 164, 181};
```
I do not understand this line

FeniksFire replied:
hahah, just copy old code and don't see it :D

alexvins commented on lib/CTownHandler.h:
```diff
@@ -253,6 +253,21 @@ class DLL_LINKAGE CTown
 
 	template <typename Handler> void serialize(Handler &h, const int version)
 	{
+		if(version < 777 && !h.saving)
```
You should not check for `h.saving` in this case. Feel free to raise exception if saving with old version but do not just ignore. `h.saving` check in serialization should be used to initialize newly added fields when loading old save.

alexvins commented on lib/CTownHandler.h:
```diff
@@ -268,15 +279,6 @@ class DLL_LINKAGE CTown
 		h & primaryRes;
 		h & warMachine;
 		h & clientInfo;
-		h & moatDamage;
```
Serialization order is important. Move backward compatibility code here

alexvins commented on lib/NetPacks.h:
```diff
@@ -1733,11 +1734,16 @@ struct ObstaclesRemoved : public CPackForClient
+			std::set<si32> obstacles; //uniqueIDs of removed obstacles
+			h & obstacles;
+		}
+		h & id;
```
this line must be conditional (only for newer version).

alexvins commented on lib/battle/RangeGenerator.cpp:
```diff
@@ -0,0 +1,50 @@
+ *
+ */
+#include "StdInc.h"
+#include "RangeGenerator.h".h"
```
Copy-past error

alexvins commented on lib/battle/obstacle/StaticObstacle.cpp:
```diff
@@ -0,0 +1,59 @@
+{
+}
+
+StaticObstacle::StaticObstacle(ObstacleJson info, int16_t position)
```
Should be `const ObstacleJson & info`

alexvins commented on lib/battle/obstacle/StaticObstacle.cpp:
```diff
@@ -0,0 +1,59 @@
+	setGraphicsInfo(info.getGraphicsInfo());
+}
+
+StaticObstacle::StaticObstacle(ObstacleJson * info, int16_t position)
```
Should be const ObstacleJson * info

alexvins commented on lib/battle/obstacle/SpellCreatedObstacle.cpp:
```diff
@@ -0,0 +1,84 @@
+	switch (getType())
+	{
+	case ObstacleType::FIRE_WALL:
+		if(spellLevel >= 2)
```
This is completely wrong. Result depends on remove obstacle spell level but not the level obstacle itself was created with.
It you want this logic be here, pass spellLevel from RemoveObstacleMechanics

alexvins commented on lib/battle/obstacle/SpellCreatedObstacle.h:
```diff
@@ -0,0 +1,44 @@
+public:
+	ObstacleType obstacleType;
+	si8 turnsRemaining;
+	si8 casterSpellPower;
```
spell power and turns remainingshould be at least 32 bit,

alexvins commented on lib/battle/obstacle/SpellCreatedObstacle.h:
```diff
@@ -0,0 +1,44 @@
+	si8 casterSpellPower;
+	si8 spellLevel;
+	si8 casterSide;
+	si8 visibleForAnotherSide;
```
this should be `bool`

alexvins commented on lib/battle/obstacle/StaticObstacle.h:
```diff
@@ -0,0 +1,35 @@
+class DLL_LINKAGE StaticObstacle : public Obstacle
+{
+public:
+	bool canBeRemovedBySpell = 0;
```
Please do not initialize boolean with a number.

alexvins commented on lib/battle/obstacle/ObstacleRandomGenerator.h:
```diff
@@ -0,0 +1,22 @@
+ *
+ */
+#pragma once
+#include "battle/RangeGenerator.h".h"
```
Copy-paste error

alexvins commented on lib/battle/obstacle/ObstacleJson.cpp:
```diff
@@ -0,0 +1,125 @@
+#include "StdInc.h"
+#include "ObstacleJson.h"
+
+ObstacleType ObstacleJson::typeConvertFromString(std::string type) const
```
should be `const std::string &`

alexvins commented on lib/battle/obstacle/ObstacleArea.h:
```diff
@@ -0,0 +1,38 @@
+ */
+#pragma once
+#include "../BattleHex.h"
+#include "boost/algorithm/searching/boyer_moore.hpp"
```
this include is not used in any of declarations, should be moved to .cpp

alexvins commented on lib/battle/obstacle/Obstacle.h:
```diff
@@ -0,0 +1,56 @@
+		h & graphicsInfo;
+	}
+private:
+	boost::uuids::uuid generateID();
```
this method should be static

alexvins commented on lib/battle/BattleInfo.cpp:
```diff
@@ -646,10 +461,174 @@ BattleInfo * BattleInfo::setupBattle(int3 tile, ETerrainType terrain, BFieldType
+			if(info.randomPosition())
+			{
+				std::vector<int> fields;
+				for(int i = 18; i<=168; i++)
```
this should be defined as constants, also it need some explanation

alexvins commented on config/obstacles.json:
```diff
@@ -3,1233 +3,1206 @@
+		"canBeRemoved" : false,
+	}
+],
+"InherentObstacles" : [
```
use camelCase for JSON field names

FeniksFire replied:
I suppose you mean lower camel case, since I used upper camel case.

alexvins commented on config/obstacles.json:
```diff
@@ -3,1233 +3,1206 @@
+		"offsetGraphicsInX" : 410,
+		"offsetGraphicsInY" : 90,
+		"graphics" : "SgCsMoat.pcx",
+		"place" : ["Castle"]
```
Where should modded castles moat configuration should be?

FeniksFire replied:
what you mean? By default moat should be in InherentObstacles since it shouldn't be random I think.

alexvins replied:
Each new town added in mod has own moat graphics

alexvins replied:
You`ve removed  moat graphics handling from `getSiegeName`, but recreated only original towns configuration in obstacles.json. What to do with modded towns?

FeniksFire replied:
Hmm, I will try with Cove but should work if you write ""place" : ["Cove"]" I suppose.

FeniksFire replied:
It's working, you can also change the graphics if you want.
![zrzut ekranu z 2017-10-14 07-57-16](https://user-images.githubusercontent.com/13856703/31573124-6304b764-b0b5-11e7-8aa4-65fcb8ae88e4.png)

FeniksFire replied:
@alexvins ok, I even extract graphics and update a hota mod for future
![zrzut ekranu z 2017-10-14 09-07-17](https://user-images.githubusercontent.com/13856703/31573591-2899559e-b0bf-11e7-94a3-496a391d4219.png)

alexvins replied:
The problem is that you can not change `config/obstacles.json` from two mods at the same time (only one override will be actually used), each mod must have own configuration. We have moat graphics format in town options (see "config/schemas/townSiege.json").

FeniksFire replied:
Old format... I deleted things that use him, since it was totally bad, it's not even about using, it's about design. I know about conflicts, there are many ways to bypass them, and back to old format is the worst. Still it's wip since there is needs to repair that conflict and random obstacle generator(now it's only generating static obstacles).

alexvins replied:
You do not have to update all the mods yourself, you "just" need to retain backward compatibility.

FeniksFire replied:
It's terrible idea for me, backward serialization is enought, I prefer update all mods, then keep that old compatibility.

alexvins replied:
Backward compatibility for mods is even more important than save format's one.

FeniksFire replied:
If all projects have such systems then nothing would develop, I can care for backwards mods compatibility but I can also update them. I can restore old code, or care and make new one, better one. Anyway I won't back to these solutions.

alexvins replied:
Lets move discussion to forum https://forum.vcmi.eu/t/poll-about-backward-compatibility/4335

alexvins commented on lib/battle/handler/BattlefieldHandler.cpp:
```diff
@@ -0,0 +1,61 @@
+
+std::vector<JsonNode> BattlefieldHandler::loadLegacyData(size_t dataSize)
+{
+	std::vector<JsonNode> data;
```
* Do not use legacy data if you actually do have it.
* even if you use legacy data API, `loadLegacyData` must return vector of individual entities, but not 1-element vector with entire configuration

alexvins commented on lib/battle/handler/BattlefieldHandler.cpp:
```diff
@@ -0,0 +1,61 @@
+BattlefieldHandler::BattlefieldHandler()
+{
+	const JsonNode obstacleConfig(JsonNode(ResourceID("config/obstacles.json")));
+
```
* why handler for obstacle types called BattlefieldHandler? Battlefield type is another thing (see `BFieldType`).
* Handler should not access .json files directly. `config/obstacles.json` link should be in `gameConfig.json` (aka mod.json for Core mod)

FeniksFire replied:
When I name it I hope that everything that is for arrangement of things and graphics on the battlefield, will be handled by this in future: siege, placement of units, graphics for battlefield etc. will go here. It's needs now a little refactor that seperate loading obstacles to another class, at all I was more worried about understand how quickly force newly handler to work to not waste too much time.

alexvins replied:
The problem is that ModHandler do not designed to support one handler for different entity types. If you plan to make Battlefield type also configurable it will also need separate handler class.

FeniksFire replied:
Maybe it was too far-reaching plan, BFieldType everything complicate, I will think about it a little more, I always change that name. However for future writing in mod.json 20 different jsons supported by 20 handlers in mod.json that set the one battle doesn't sound too good. It would be better to have one json that can set whole battle.

alexvins commented on lib/battle/handler/BattlefieldHandler.cpp:
```diff
@@ -0,0 +1,61 @@
+{
+	std::vector<bool> allowed;
+	for(auto i : obstacleConfigs)
+		allowed.push_back(true);
```
Use `allowed.resize(obstacleConfigs.size(), true)`

alexvins commented on lib/battle/handler/BattlefieldHandler.cpp:
```diff
@@ -0,0 +1,61 @@
+
+std::vector<std::shared_ptr<ObstacleJson> > BattlefieldHandler::getObstacleConfigs() const
+{
+	return obstacleConfigs;
```
You`re should not copy entire vector here.

alexvins commented on lib/battle/handler/BattlefieldHandler.h:
```diff
@@ -0,0 +1,33 @@
+
+class ObstacleJson;
+class DLL_LINKAGE BattlefieldHandler : public IHandlerBase
+{
```
Handler should be serialized as part of LibClasses.

alexvins commented on lib/battle/obstacle/ObstacleJson.cpp:
```diff
@@ -114,6 +114,13 @@ bool ObstacleJson::randomPosition() const
+{
+	if(config["inherent"].isNull())
+		return false;
+	return config["inherent"].Bool();
```
This line is enough. `Bool()` returns `false` for null node.

alexvins commented on config/obstacles.json:
```diff
@@ -20,1189 +20,1346 @@
-		"place" : ["Tower", "Crypt"]
-	}
-]
+"Obstacle001" :
```
Object identifiers should also be in lowerCamelCase.

alexvins reviewed (dismissed):
This one is postponed until #359 merged

Commit:
Minor fixes and start work on additional animations

-Battlefield type is now randomize properly
-Adding equal operators to UUID
-ObstacleGraphicsInfo can store two more names for animations

alexvins commented on lib/battle/obstacle/ObstacleGraphicsInfo.h:
```diff
@@ -12,31 +12,35 @@
 		h & offsetGraphicsInX;
 		h & offsetGraphicsInY;
-		h & graphicsName;
+		h & graphics[0];
```
Serializer can handle arrays by itself

alexvins commented on client/battle/CBattleInterface.cpp:
```diff
@@ -3401,15 +3290,29 @@ void CBattleInterface::showStacks(SDL_Surface *to, std::vector<const CStack *> s
 		{
-			Point p = getObstaclePosition(img, *obstacle);
-			img->draw(to, p.x, p.y);
+			auto bitmap = BitmapHandler::loadBitmap(obstacle->getGraphicsInfo().getGraphics(obstacle->state));
```
Same here, try to cache obstacle bitmap.

alexvins commented on client/battle/CBattleInterface.cpp:
```diff
@@ -3752,3 +3589,30 @@ void CBattleInterface::showPiecesOfWall(SDL_Surface *to, std::vector<int> pieces
 	}
 }
+
+void CBattleInterface::obstacleAnimEffect(const Obstacle &oi, ObstacleState state, bool waitForAnimations)
```
- Missing space after `&`. 
- Is ObstacleState copying intended?

alexvins commented on client/battle/CBattleInterface.cpp:
```diff
@@ -3752,3 +3589,30 @@ void CBattleInterface::showPiecesOfWall(SDL_Surface *to, std::vector<int> pieces
+		CAnimation anim(graphicsname);
+		anim.preload();
+		int frameIndex = ((animCount + 1) * 25 / getAnimSpeed()) % anim.size();
+		imageHeight = anim.getImage(frameIndex)->height();
```
You should find more effective way of obtaining height.

alexvins commented on client/battle/CBattleInterface.h:
```diff
@@ -1,407 +1,404 @@
-/*
```
Whole file changed, line ending problem?

FeniksFire replied:
I made small test, by using push -f I checked for LF, CRLF, CR, but it's not resolve the problem, any ideas what it can be?

alexvins commented on client/widgets/TextControls.h:
```diff
@@ -42,7 +42,7 @@ class CLabel : public CTextContainer
 	std::shared_ptr<CPicture> background;
 public:
-
+	
```
just revert this

alexvins commented on lib/CHeroHandler.h:
```diff
@@ -329,7 +320,12 @@ class DLL_LINKAGE CHeroHandler : public IHandlerBase
+		{
+			std::map<int, CObstacleInfo> obstacles;
+			std::map<int, CObstacleInfo> absoluteObstacles;
 		h & obstacles;
```
Fix indentation here.

alexvins commented on lib/CMakeLists.txt:
```diff
@@ -346,10 +377,11 @@ set(lib_HEADERS
 		vcmi_endian.h
 		VCMI_Lib.h
-)
+        battle/obstacle/ObstacleState.h)
```
Put this line near other battle/obstacle headers

alexvins commented on lib/CModHandler.cpp:
```diff
@@ -17,6 +17,7 @@
 #include "CCreatureHandler.h"
 #include "CArtHandler.h"
 #include "CTownHandler.h"
+#include "battle/handler/BattlefieldHandler.h"
```
use relative path

FeniksFire replied:
What is wrong with it?

alexvins replied:
Now it looks OK )

alexvins commented on lib/NetPacksLib.cpp:
```diff
@@ -1,1622 +1,1625 @@
-/*
```
Whole file changed, problem with line endings?

FeniksFire replied:
I made small test, by using push -f I checked for LF, CRLF, CR, but it's not resolve the problem, any ideas what it can be?

alexvins commented on lib/VCMI_Lib.h:
```diff
@@ -66,6 +68,7 @@ class DLL_LINKAGE LibClasses
 		h & arth;
 		h & creh;
 		h & townh;
+		h & battlefieldHandler;
```
Backward compatibility code is missing.

alexvins commented on lib/battle/BattleInfo.cpp:
```diff
@@ -1,1093 +1,1035 @@
-/*
```
Whole file changed, problem with line endings?

FeniksFire replied:
I made small test, by using push -f I checked for LF, CRLF, CR, but it's not resolve the problem, any ideas what it can be?

alexvins commented on lib/battle/BattleInfo.h:
```diff
@@ -1,151 +1,154 @@
-/*
```
Whole file changed, problem with line endings?

FeniksFire replied:
I made small test, by using push -f I checked for LF, CRLF, CR, but it's not resolve the problem, any ideas what it can be?

alexvins replied:
set 
``` 
[core]
	safecrlf = warn
```
in .gitconfig and recommit ff3a2148

FeniksFire replied:
Thanks! Will check tommorow.

FeniksFire replied:
Well with `safectrlf = warn` each time when I want  commit I get `fatal: bad numeric config value 'warn' for 'core.autocrlf': invalid unit`
Are you sure there must be `warn`?

alexvins replied:
It should work, see https://git-scm.com/docs/git-config#git-config-coresafecrlf

FeniksFire replied:
Huh, there were 2 problems. 
1) CLion seems to have some bug since it doesn't change line endings.
2)You was close but instead of `git config --global core.autocrlf warn` change to `git config --global core.autocrlf false` and this help me a bit.

alexvins commented on lib/battle/obstacle/Obstacle.h:
```diff
@@ -0,0 +1,58 @@
+			h & state;
+			h & pos;
+		}
+		h & ID;
```
This should not be called in case of `version < 777`

alexvins commented on lib/registerTypes/RegisterTypes.h:
```diff
@@ -208,9 +209,9 @@ void registerTypesMapObjects2(Serializer &s)
-	//s.template registerType<CObstacleInstance>();
-		s.template registerType<CObstacleInstance, MoatObstacle>();
-		s.template registerType<CObstacleInstance, SpellCreatedObstacle>();
+	s.template registerType<Obstacle, StaticObstacle>();
```
This is a problem. You registering 4 classes where were 3. This completely breaks saves compatibilty.

Commit:
Obstacle states

-Start making obstacle to handle 4 states: Appear, Default, Interaction, Disappear
-Each states can have own animation

alexvins commented on lib/UUID.cpp:
```diff
@@ -0,0 +1,33 @@
+
+boost::uuids::uuid UUID::generateID()
+{
+	static boost::uuids::random_generator_mt19937 gen;
```
Build fails here. No such symbol `random_generator_mt19937`. Boost version problem?

FeniksFire replied:
You're right, CI works with `boost 1.64` when `boost 1.67` provide for  convenience 
` random_generator_mt19937` that is definition of previous version  `basic_random_generator<mt19937>`

alexvins commented on lib/CHeroHandler.h:
```diff
@@ -250,6 +222,29 @@ class DLL_LINKAGE CHeroClassHandler : public IHandlerBase
 	}
 };
 
+struct CObstacleInfo // for backward compatibility
```
Put this where old CObstacleInfo was

FeniksFire replied:
Ahh, ok.

Commit:
Fixes

-Unit tests
